File: examples/go/tests/json_schema_chat_test.go
package tests

import (
	"testing"

	apichat "examples/go/json_schema/api/chat"

	"github.com/stretchr/testify/require"
	"go.temporal.io/sdk/converter"
	"go.temporal.io/sdk/workflow"
)

// TestJSONSchemaChatRuntime round-trips every chat wire fixture through the
// Temporal default data converter and asserts JSON-equality against the
// canonical fixtures, mirroring the Python and Java suites.
//
// Exception (see json-schema/nullability.md): optional+nullable fields collapse
// in Go. message-full.json carries replyToId: null (optional+nullable), which
// Go collapses on serialize, so it is verified by deserialization + field checks
// rather than exact JSON-equality, matching the Java test. (room-open.json's
// topic is required-nullable, so its explicit null survives the round-trip and
// is checked via JSON-equality.)
func TestJSONSchemaChatRuntime(t *testing.T) {
	dc := converter.GetDefaultDataConverter()

	minimal := roundTripJSONEq[apichat.Message](t, dc, "chat", "message-minimal.json")
	require.Equal(t, apichat.MessageKindText, minimal.Kind)
	require.Equal(t, "hi", minimal.Body)
	require.Nil(t, minimal.ReplyToId)
	require.Nil(t, minimal.Priority)
	require.Equal(t, int64(0), minimal.PriorityOrDefault())

	// message-full carries replyToId: null (optional+nullable) — deserialization only.
	full := decodeFixture[apichat.Message](t, dc, "chat", "message-full.json")
	require.Nil(t, full.ReplyToId)
	require.NotNil(t, full.Priority)
	require.Equal(t, int64(7), *full.Priority)

	room := roundTripJSONEq[apichat.Room](t, dc, "chat", "room-open.json")
	require.Equal(t, "r1", room.RoomId)
	require.Nil(t, room.Topic)
	require.Equal(t, []string{"a"}, room.Members)
	require.Contains(t, room.AdditionalProperties, "x-extra")

	labels := roundTripJSONEq[apichat.Labels](t, dc, "chat", "labels.json")
	require.Equal(t, "prod", labels.AdditionalProperties["env"])
	require.Equal(t, "core", labels.AdditionalProperties["team"])

	input := roundTripJSONEq[apichat.SendMessageInput](t, dc, "chat", "send-message-input.json")
	require.Equal(t, "r1", input.RoomId)
	require.Equal(t, "hi", input.Message.Body)

	output := roundTripJSONEq[apichat.SendMessageOutput](t, dc, "chat", "send-message-output.json")
	require.Equal(t, "m1", output.MessageId)

	require.Equal(t, "example.chat.v1.ChatService", apichat.ChatService.ServiceName)
	require.Equal(t, "SendMessage", apichat.ChatService.SendMessage.Name())
	require.NotNil(t, apichat.NewChatServiceClient("chat-endpoint"))
}

var _ func(*apichat.ChatServiceClient, workflow.Context, apichat.SendMessageInput) workflow.Future = (*apichat.ChatServiceClient).SendMessage
var _ func(*apichat.ChatServiceClient, workflow.Context, apichat.GetRoomInput) workflow.Future = (*apichat.ChatServiceClient).GetRoom
var _ func(*apichat.ChatServiceClient, workflow.Context) workflow.Future = (*apichat.ChatServiceClient).Ping

File: examples/typescript/tests/json-schema-chat.test.ts
import { describe, expect, test } from "vitest";

import {
  DEFAULT_PRIORITY,
  LabelsMapper,
  MessageMapper,
  RoomMapper,
  SendMessageInputMapper,
  SendMessageOutputMapper,
  type Labels,
  type Message,
} from "../json_schema/definitions/chat/models.ts";
import { ValidationError } from "../json_schema/definitions/chat/definitions.ts";
import {
  fixtureBytes,
  loadFixture as loadFixtureFrom,
  roundTripFixture,
  type IntermediateMapper,
} from "./json-converter-helper.ts";

const wireFixtureDir = new URL("../../wire/json_schema/chat/", import.meta.url);

function loadFixture(name: string): unknown {
  return loadFixtureFrom(wireFixtureDir, name);
}

// Round-trip a fixture through the Temporal data converter (driven by the
// generated mapper) and assert the re-serialized JSON is JSON-equal to the
// fixture. TS mappers preserve explicit nulls, so all chat fixtures use exact
// JSON-equality (no optional+nullable collapse — unlike Go).
function expectRoundTrip<T>(name: string, mapper: IntermediateMapper<T>): T {
  const { value, serialized } = roundTripFixture(
    mapper,
    fixtureBytes(wireFixtureDir, name),
  );
  expect(serialized).toEqual(loadFixture(name));
  return value;
}

describe("json-schema chat generated definitions", () => {
  test("roundtrips canonical wire fixtures through the Temporal converter", () => {
    const message = expectRoundTrip("message-minimal.json", new MessageMapper());
    expect(message).toMatchObject<Message>({
      kind: "text",
      body: "hi",
    });
    expect(message.replyToId).toBeUndefined();
    expect(message.priority ?? DEFAULT_PRIORITY).toBe(0);

    const fullMessage = expectRoundTrip("message-full.json", new MessageMapper());
    expect(fullMessage.replyToId).toBeNull();
    expect(fullMessage.priority).toBe(7);

    const room = expectRoundTrip("room-open.json", new RoomMapper());
    expect(room.additionalProperties).toEqual({ "x-extra": 42 });

    const labels = expectRoundTrip("labels.json", new LabelsMapper());
    expect(labels).toMatchObject<Labels>({
      additionalProperties: { env: "prod", team: "core" },
    });

    const request = expectRoundTrip(
      "send-message-input.json",
      new SendMessageInputMapper(),
    );
    expect(request.message.body).toBe("hi");

    const response = expectRoundTrip(
      "send-message-output.json",
      new SendMessageOutputMapper(),
    );
    expect(response.messageId).toBe("m1");
  });

  test("reports JSON schema validation errors", () => {
    expect(() =>
      new SendMessageInputMapper().fromIntermediate({
        roomId: "r1",
        message: { kind: "text", body: "hi" },
        extra: true,
      }),
    ).toThrow(ValidationError);

    expect(() =>
      new MessageMapper().fromIntermediate({ kind: "image", body: "hi" }),
    ).toThrow(ValidationError);

    expect(() => new SendMessageOutputMapper().fromIntermediate({})).toThrow(
      ValidationError,
    );
  });
});
